README.md
## La Javaness Annotto API
[![Test](https://github.com/lajavaness/annotto-api/actions/workflows/tests.yml/badge.svg)](https://github.com/lajavaness/annotto-api/actions/workflows/tests.yml)

Is the API to create a new project and store a corresponding dataset and then annotate data.

## Prerequisites
1. Node 16+
2. Mongo 4.4
3. Keycloak 15.0.1
4. PostgreSQL 13+ (as Keycloak DB)

## Launch the app

Install project dependencies
```shell
yarn install
```
Launch environment
```shell
docker-compose up -d
```
Launch annotto-api

### Environment variables
| Name                         | Default                                   | Optional-Required | Description                                                          |
|------------------------------|-------------------------------------------|---------------|----------------------------------------------------------------------|
| PORT                         | 5001                                      | optional      | Server listening port                                                |
| NODE_ENV                     | development                               | optional      | NODE Environment to use "[development, test]"                        |
| ENCRYPTION_SECRET_KEY        | -                                         | optional      | A Secret Key used to encrypt AWS creds  (symmetric)                  |
| MONGO_URL                    | mongodb://localhost:27017/ljn_annotto_dev | optional      | Mongo connection string                                              |
| ANNOTTO_FRONT_URL            | http://localhost:3000                     | optional      | Annotto Front base url                                               |
| KEYCLOAK_REALM               | annotto                                   | optional      | Keycloak Realm (preconfigured if started with docker-compose_)       |
| KEYCLOAK_AUTH_URL            | http://localhost:8080/auth                | optional      | Keycloak auth url (preconfigured if started with docker-compose_)    |
| KEYCLOAK_CLIENT_ID           | annotto                                   | optional      | Keycloak client id (preconfigured if started with docker-compose_)   |
| KEYCLOAK_CLIENT_SECRET       | a7b7a29d-abb0-4e21-abec-bca99a47e40e      | optional      | Keycloak client secret (preconfigured if started with docker-compose_) |
| ANNOTTO_UPLOAD_MAX_FILE_SIZE | 1048576000                                | optional      | Max file size permitted to upload (default = 1000 * 1024 * 1024)     |
| ANNOTTO_UPLOAD_BATCH_SIZE    | 50000                                     | optional      | Max file size permitted to upload (default = 1000 * 1024 * 1024)     |


```shell
yarn start
```

You should have this running now
1. Annotto API (http://localhost:5001).
   * Swagger available at http://localhost:5001/api-docs
   * Health status available at http://localhost:5001/health
2. Keycloak
   * Admin Console:
      - url: http://localhost:8080
      - username: admin
      - password: admin

### Configuration
The configuration resides in the `config` folder. Two configuration mode is available
1. development (`config/development.js`)
2. test (`config/test.js`)

You can override the default config in `config/index.js` with those specific files.

## Keycloak
Annotto is protected by OAuth using Keycloak. When you start the environment with docker-compose, you will automatically 
create users to be able to make authorized requests to annotto API.

The users automatically created are:
|username|password|role|
|---|---|---|
|admin|test|Admin|
|data|test|Data Scientist|
|user|test|User|

By default, an OAuth Realm `annotto` with basics users and roles are preconfigured when launching with docker-compose.yml
Once the service is started, you can manage your Keycloak instance at your will.

You can see more documentation on Keycloak and Annotto [here](./docs/keycloak.md)


##  Contributing

### Team Conventions

This package is co-maintained by all front-end developers at [La Javaness](https://www.lajavaness.com). Each contribution must follow established conventions.

We ensure consistent code formatting with [Prettier](https://prettier.io), code quality with [ESLint](https://eslint.org), and commit message formatting with [commitlint](https://commitlint.js.org).

Incoming code must not cause existing tests to fail; that rule is enforced with a precommit hook.

Code should be pushed to a ``feature/*`` or ``fix/*`` branch based on the type of change made, and a Pull Request made against the next branch. A fellow developer must approve the PR before it can be merged.

### Commits

#### Formatting
The project uses [semantic-release](https://semantic-release.gitbook.io/semantic-release) to upgrade, document and deploy its NPM package, and [commitlint](https://commitlint.js.org) to ensure all commit messages are understood by semantic-release.

Please read the [upstream documentation](https://semantic-release.gitbook.io/semantic-release/#commit-message-format) to understand the enforced format. Below is an excerpt for quick reference:

| Commit message                                                     | Release type             |
|:-------------------------------------------------------------------|:-------------------------|
| fix(pencil): Stop graphite breaking when too much pressure applied | Patch Release            |
| docs(README): Append a `Use graphite properly` section to README   | Patch Release            |
| refactor: Improve 'graphiteWidth' option                           | Patch Release            |
| style: Remove all semi-colons                                      | Patch Release            |
| feat(pencil): Add 'graphiteWidth' option                           | (Minor) Feature Release  |
| perf(pencil): Remove graphiteWidth option                          | (Major) Breaking Release |
| BREAKING CHANGE: The graphiteWidth option has been removed.        | (Major) Breaking Release |

Below is the meaning of the different commit prefixes:

- ``feat``: (new feature for a component or new component)
- ``fix``: (bug fix in a component)
- ``docs``: (changes to the documentation)
- ``style``: (formatting, missing semi colons, etc; no production code change)
- ``refactor``: (refactoring production code, eg. renaming a variable)
- ``test``: (adding missing tests, refactoring tests; no code change)
- ``perf``: (changes to code for performance reasons)
- ``chore``: (updating yarn tasks, CI pipelines, build scripts, etc; no code change)

#### Precommit Hooks

Both ``test:ci`` and ``prettier`` commands are run before each commit to ensure the code remains well-formatted and tested.

Ensure the `.git/hooks/pre-commit`` file is present in your repository.

The ``yarn prettier`` command used by pre-commit hooks automatically fixes style issues and stages any unstaged content into your commit, to make sure those style changes are staged. This prevents the use of partial commits. If you want to commit only part of your changed files, you can ``stash`` unstaged code before committing. This is the preferred way of doing things at La Javaness.




CHANGELOG.md
## [1.0.21](https://github.com/lajavaness/annotto-api/compare/v1.0.20...v1.0.21) (2022-07-12)

## [1.0.20](https://github.com/lajavaness/annotto-api/compare/v1.0.19...v1.0.20) (2022-07-12)

## [1.0.19](https://github.com/lajavaness/annotto-api/compare/v1.0.18...v1.0.19) (2022-07-11)

## [1.0.18](https://github.com/lajavaness/annotto-api/compare/v1.0.17...v1.0.18) (2022-06-29)

## [1.0.17](https://github.com/lajavaness/annotto-api/compare/v1.0.16...v1.0.17) (2022-06-29)

## [1.0.16](https://github.com/lajavaness/annotto-api/compare/v1.0.15...v1.0.16) (2022-06-29)

## [1.0.15](https://github.com/lajavaness/annotto-api/compare/v1.0.14...v1.0.15) (2022-06-29)

## [1.0.14](https://github.com/lajavaness/annotto-api/compare/v1.0.13...v1.0.14) (2022-06-29)

## [1.0.13](https://github.com/lajavaness/annotto-api/compare/v1.0.12...v1.0.13) (2022-06-29)

## [1.0.12](https://github.com/lajavaness/annotto-api/compare/v1.0.11...v1.0.12) (2022-06-29)

## [1.0.11](https://github.com/lajavaness/annotto-api/compare/v1.0.10...v1.0.11) (2022-06-29)

## [1.0.10](https://github.com/lajavaness/annotto-api/compare/v1.0.9...v1.0.10) (2022-06-29)

## [1.0.9](https://github.com/lajavaness/annotto-api/compare/v1.0.8...v1.0.9) (2022-06-29)

## [1.0.8](https://github.com/lajavaness/annotto-api/compare/v1.0.7...v1.0.8) (2022-06-29)

## [1.0.7](https://github.com/lajavaness/annotto-api/compare/v1.0.6...v1.0.7) (2022-06-29)

## [1.0.6](https://github.com/lajavaness/annotto-api/compare/v1.0.5...v1.0.6) (2022-06-29)

## [1.0.5](https://github.com/lajavaness/annotto-api/compare/v1.0.4...v1.0.5) (2022-06-29)

## [1.0.4](https://github.com/lajavaness/annotto-api/compare/v1.0.3...v1.0.4) (2022-06-29)

## [1.0.3](https://github.com/lajavaness/annotto-api/compare/v1.0.2...v1.0.3) (2022-06-29)

## [1.0.2](https://github.com/lajavaness/annotto-api/compare/v1.0.1...v1.0.2) (2022-06-23)

## [1.0.1](https://github.com/lajavaness/annotto-api/compare/v1.0.0...v1.0.1) (2022-06-22)


### Bug Fixes

* Fix undefined task properties in project config, [#1](https://github.com/lajavaness/annotto-api/issues/1) ([f84713e](https://github.com/lajavaness/annotto-api/commit/f84713e4724dc2cc65c7fb837ad6674930ca819a))

# 1.0.0 (2022-06-16)


### Features

* Initial commit ([7a95b56](https://github.com/lajavaness/annotto-api/commit/7a95b5669e2b5442e585cf7c9b3583e63fae0b10))


config/config.js
const LIMIT = 10

module.exports = {
  demo: true, // Enable this to create demo projects at startup
  port: process.env.PORT || 5001,
  encryptSecretKey: process.env.ENCRYPTION_SECRET_KEY, // Encryption is used typically when providing AWS access key and secret key
  mongo: {
    url: process.env.MONGO_URL || 'mongodb://localhost:27017/ljn_annotto_dev',
    options: {
      useCreateIndex: true,
      useNewUrlParser: true,
      useUnifiedTopology: true,
      useFindAndModify: false,
    },
  },
  cors: {
    exposedHeaders: ['Content-Length', 'X-Requested-With', 'X-Total-Count'],
    credentials: true,
    origin: process.env.ANNOTTO_FRONT_URL || 'http://localhost:3000',
  },
  keycloak: {
    realm: process.env.KEYCLOAK_REALM || 'annotto',
    'auth-server-url': process.env.KEYCLOAK_AUTH_URL || 'http://localhost:8080/auth',
    resource: process.env.KEYCLOAK_CLIENT_ID || 'annotto',
    clientSecret: process.env.KEYCLOAK_CLIENT_SECRET || 'a7b7a29d-abb0-4e21-abec-bca99a47e40e',
  },
  swagger: {
    swaggerUi: '/explorer',
    apiDocs: '/api-docs',
  },
  baseUrl: 'http://127.0.0.1:5001',
  fileUpload: {
    maxFileSize: process.env.ANNOTTO_UPLOAD_MAX_FILE_SIZE
      ? parseInt(process.env.ANNOTTO_UPLOAD_MAX_FILE_SIZE)
      : 1000 * 1024 * 1024,
    batchSize: process.env.ANNOTTO_UPLOAD_BATCH_SIZE ? parseInt(process.env.ANNOTTO_UPLOAD_BATCH_SIZE) : 50000,
  },
  search: {
    tasks: {
      orderBy: ['label'],
      fields: {
        classificationId: { key: '_id', type: 'number' },
        projectId: { key: 'project', type: 'number' },
      },
      limit: LIMIT,
      select: {
        annotationCount: true,
        annotationPourcent: true,
        category: true,
        value: true,
        type: true,
        label: true,
        min: true,
        max: true,
      },
    },
    user: {
      orderBy: ['lastName'],
      fields: {
        idUser: { key: '_id', type: 'number' },
        firstName: { key: 'firstName', type: 'text' },
        lastName: { key: 'lastName', type: 'text' },
        email: { key: 'email', type: 'text' },
      },
      limit: LIMIT,
    },
    client: {
      orderBy: ['name'],
      fields: {
        clientId: { key: '_id', type: 'number' },
        name: { key: 'name', type: 'text' },
        isActive: { key: 'isActive', type: 'boolean' },
        description: { key: 'description', type: 'text' },
      },
      limit: LIMIT,
    },
    project: {
      orderBy: ['name'],
      fields: {
        projectId: { key: '_id', type: 'number' },
        clientId: { key: 'client', type: 'string' },
        name: { key: 'name', type: 'text' },
        active: { key: 'active', type: 'boolean' },
        description: { key: 'description', type: 'text' },
      },
      limit: LIMIT,
      select: {
        client: true,
        admins: true,
        users: true,
        dataScientists: true,
        itemCount: true,
        commentCount: true,
        deadline: true,
        progress: true,
        velocity: true,
        remainingWork: true,
        lastAnnotationTime: true,
        name: true,
        updatedAt: true,
      },
    },
    batch: {
      orderBy: ['name'],
      fields: {
        name: { key: 'name', type: 'string' },
        description: { key: 'description', type: 'text' },
        projectId: { key: 'project', type: 'string' },
      },
      limit: LIMIT,
    },
    item: {
      orderBy: ['updatedAt'],
      fields: {
        projectId: { key: 'project', type: 'number' },
        status: { key: 'status', type: 'string' },
        itemId: { key: '_id', type: 'number' },
        type: { key: 'type', type: 'string' },
        body: { key: 'body', type: 'text' },
        tags: { key: 'tags', type: 'array' },
        annotated: { key: 'annotated', type: 'boolean' },
        uuid: { key: 'uuid', type: 'string' },
        compositeUuid: { key: 'compositeUuid', type: 'string' },
        updatedAt: { key: 'updatedAt', type: 'string' },
      },
      limit: LIMIT,
      select: {
        tags: true,
        commentCount: true,
        logCount: true,
        lastAnnotator: true,
        annotationValues: true,
        annotatedAt: true,
        velocity: true,
        body: true,
        annotated: true,
      },
    },
    comment: {
      orderBy: ['-createdAt'],
      fields: {
        comment: { key: 'comment', type: 'string' },
        itemId: { key: 'item', type: 'string' },
        projectId: { key: 'project', type: 'string' },
        batchId: { key: 'batch', type: 'string' },
        userId: { key: 'user', type: 'string' },
        createdAt: { key: 'createdAt', type: 'string' },
      },
      limit: LIMIT,
    },
    log: {
      orderBy: ['-createdAt'],
      fields: {
        comment: { key: 'comment', type: 'string' },
        itemId: { key: 'item', type: 'string' },
        type: { key: 'type', type: 'text' },
        projectId: { key: 'project', type: 'string' },
        batchId: { key: 'batch', type: 'string' },
        userId: { key: 'user', type: 'string' },
        commentType: { key: 'commentType', type: 'string' },
        projectType: { key: 'projectType', type: 'string' },
        missionType: { key: 'missionType', type: 'string' },
        createdAt: { key: 'createdAt', type: 'string' },
      },
      limit: LIMIT,
    },
  },
  filter: {
    items: {
      operators: [
        {
          name: 'equal',
          param: { value: 'Boolean' },
        },
        {
          name: 'range',
          param: { value: { from: 'Date', to: 'Date' } },
        },
        {
          name: 'containsAny',
          param: { value: ['String'] },
        },
        {
          name: 'containsAll',
          param: { value: ['String'] },
        },
        {
          name: 'size',
          param: { value: 'Number' },
        },
        {
          name: 'similarTo',
          param: { value: 'String' },
          optionalParam: {
            limit: 'Number',
            neg_values: ['String'],
          },
        },
        {
          name: 'greaterThanAny',
          param: {
            value: ['String'],
          },
          optionalParam: {
            threshold: 'Number',
          },
        },
        {
          name: 'greaterThanAll',
          param: {
            value: ['String'],
          },
          optionalParam: {
            threshold: 'Number',
          },
        },
        {
          name: 'wrongPredictions',
          param: {
            value: ['String'],
          },
          optionalParam: {
            threshold: 'Number',
          },
        },
        {
          name: 'textContains',
          param: {
            value: 'String',
          },
        },
      ],
      fields: {
        annotated: { key: 'annotated', operators: ['equal'] },
        annotatedAt: { key: 'annotatedAt', operators: ['range'] },
        annotatedBy: { key: 'annotatedBy', operators: ['containsAny', 'containsAll'] },
        annotationValues: { key: 'annotationValues', operators: ['containsAny', 'containsAll', 'size'] },
        uuid: { key: 'compositeUuid', operators: ['similarTo'] },
        tags: { key: 'tags', operators: ['containsAny', 'containsAll', 'size'] },
        predictionValues: { key: 'predictions.keys', operators: ['containsAny', 'containsAll', 'size'] },
        predictionScores: {
          key: 'predictions.predictions',
          operators: ['greaterThanAny', 'greaterThanAll', 'wrongPredictions'],
        },
        body: { key: 'body', operators: ['textContains'] },
      },
    },
  },
}
